### PR TITLE
Feature: Package owner badge + metadata

### PR DESCRIPTION
**Pull request by talekhinezh**
* Add Package owner badge which is returned when package is created
* Add Package metadata which may be updated by owner of owner badge
* Simplified native invocation architecture

### CODE REVIEW

File: radix-engine/src/engine/interpreters/native_interpreter.rs [hunk 1]
```diff
@@ -76,77 +81,21 @@ impl Into<ApplicationError> for EpochManagerError {
...
-}
-
-pub trait NativeExecutable: Invocation {
-    type NativeOutput: Debug;
-
-    fn execute<Y>(
-        invocation: Self,
-        system_api: &mut Y,
-    ) -> Result<(<Self as Invocation>::Output, CallFrameUpdate), RuntimeError>
+pub trait NativeProgram {
```
Comment (iamyulong) [inline radix-engine/src/engine/interpreters/native_interpreter.rs:84]: Program sounds a bit heavy. Suggest:
```rust
NativeProcedure {
    fn run()
}
```

File: radix-engine/src/engine/modules/auth_module.rs [hunk 2]
```diff
@@ -84,25 +85,76 @@ impl AuthModule {
...
                 match (method, resolved_receiver) {
+                    (ResolvedMethod::Native(NativeMethod::Metadata(MetadataMethod::Set)), ..) => {
+                        if let Some((
+                            RENodeId::Global(GlobalAddress::Package(package_address)),
+                            ..,
+                        )) = resolved_receiver.derefed_from
+                        {
+                            let bytes = scrypto_encode(&package_address).map_err(|_| {
+                                RuntimeError::ModuleError(ModuleError::AuthError(
+                                    AuthError::CouldNotEncodePackageAddress,
```
Comment (iamyulong) [inline radix-engine/src/engine/modules/auth_module.rs:96]: Seems like we should panic.

File: radix-engine/src/engine/native_wrapper.rs [hunk 3]
```diff
@@ -47,13 +47,22 @@ where
...
             }
-            NativeFunction::Package(PackageFunction::Publish) => {
-                let invocation: PackagePublishInvocation = scrypto_decode(&args)
-                    .map_err(|e| RuntimeError::KernelError(KernelError::InvalidSborValue(e)))?;
-                system_api
-                    .sys_invoke(invocation)
-                    .map(|a| IndexedScryptoValue::from_typed(&a))
-            }
+            NativeFunction::Package(package_function) => match package_function {
+                PackageFunction::PublishNoOwner => {
```
Comment (iamyulong) [inline radix-engine/src/engine/native_wrapper.rs:51]: Same

File: radix-engine-interface/src/wasm/input.rs [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+pub enum NativeFnInvocation {
+    Method(NativeMethodInvocation),
+    Function(NativeFunctionInvocation),
+}
+
+#[derive(Debug)]
+#[scrypto(TypeId, Encode, Decode)]
+pub enum NativeMethodInvocation {
+    AccessRules(AccessRulesMethodInvocation),
+    Metadata(MetadataMethodInvocation),
```
Comment (iamyulong) [inline radix-engine-interface/src/wasm/input.rs:42]: I think, at some point, we will need to limit and isolate these "any" methods. My inclination is to make it:
```rust
pub enum NativeMethodInvocation {
    Global(GlobalMethodInvocation)
}
```

**Review by iamyulong — COMMENTED** [review]
I started to worry that mixing generic substates (i.e. access rule and metadata) with type substatess will cause more pain in the long run. This is essentially unneeded flexibilty.

Would very much like to explore if we should limit them to just globals from a practical viewpoint.

File: radix-engine-interface/src/api/types/invocation/native.rs [hunk 5]
```diff
@@ -322,7 +344,8 @@ pub enum WorktopMethod {
 #[scrypto(TypeId, Encode, Decode, Describe)]
 #[strum(serialize_all = "snake_case")]
 pub enum PackageFunction {
-    Publish,
+    PublishNoOwner,
+    PublishWithOwner,
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine-interface/src/api/types/invocation/native.rs:348]: Will add a note here regarding NoOwner

File: radix-engine/src/engine/interpreters/native_interpreter.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/engine/interpreters/native_interpreter.rs:84]: sure sounds good

File: radix-engine/src/engine/modules/auth_module.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/src/engine/modules/auth_module.rs:96]: okay

**Comment by iamyulong** [comment]
Discussed offline. 

Thoughts on trait methods have been documented on https://github.com/radixdlt/radixdlt-scrypto/pull/599.

Will keep the "NoOwner" variant for now, otherwise we will have to find another place to store the system entity owner badge.  Josh may challenge the decision of not having NoOwner again. :)

**Review by iamyulong — APPROVED** [review]

**Pull request merged**